DT-32 Histogram of clicks and impressions

vanesssalai opened the pull request:
Added query for clicks and impressions for mentors (grouped by industry). currently displayed as a table with the top and bottom 5 in the page 'Clicks & Impressions'

wei2912 commented on dashboards/clicks_and_impressions.py:
```diff
@@ -0,0 +1,64 @@
+df_processed["query_params"] = df_processed["url"].apply(extract_query_params)
+df_processed["parsed_params"] = df_processed["query_params"].apply(process_query_params)
+
+# helper to extract industry
+def extract_industry(params):
+    industries = params.get("industries", [])
+    if isinstance(industries, list) and industries:
+        return industries[0]
+    return None
+
+df_processed["industries"] = df_processed["parsed_params"].apply(extract_industry)
```
From what I understand of this code, a click/impression on a mentor is considered to be of Industry X if the search query parameter includes Industry X. However, this isn't necessarily the case - in fact, the URL visit isn't sufficient and one needs to use the Elasticsearch data on individual mentors in order to obtain their industry (see https://github.com/AdvisorySG/mentorship-streamlit/issues/42).

wei2912 commented on dashboards/clicks_and_impressions.py:
```diff
@@ -0,0 +1,64 @@
+# Initialize connection
+conn = st.connection("mysql", type="sql")
+
+df = conn.query("select * from website_event;")
+
+df_processed = df.copy(deep=True)
+#
+df_processed["url"] = "/?" + df["url_query"].astype(str)
+df_processed["query_params"] = df_processed["url"].apply(extract_query_params)
+df_processed["parsed_params"] = df_processed["query_params"].apply(process_query_params)
```
Since https://github.com/AdvisorySG/mentorship-streamlit/commit/9ef75bd5763d94f7d964c1bb48824a295782f00c has been merged, please use DuckDB SQL queries instead to retrieve the initial dataframe. Some sample code is available at https://github.com/AdvisorySG/mentorship-streamlit/issues/48.

wei2912 reviewed (changes requested):
Grouping by industries is useful, but would require some additional code to take in Elasticsearch data and then perform a JOIN query to match the recorded document IDs with the mentor profiles.

For now, I would like to just have the following implemented:

1. Filter only production events

Each tracked event has an unique ID `event_data_id` with two fields, `data_key` and `string_value` (for now the other `..._value` fields aren't used). Each pair corresponds to a single row in the table:

![image](https://github.com/user-attachments/assets/a7d451c4-b55d-4c05-a734-8c8c92e854b3)

Filtering by `data_key = 'production'` is required to remove development events from the database.

2. Calculate number of clicks and impressions on a per-record basis

3. Histogram for these clicks/impressions (to see if there are particular mentors that receive a very high no. of clicks/impressions)

vanesssalai commented:
As mentioned above, I updated the query to use DuckDB SQL it now will group by mentor_name as the id field will create 2 different rows for some mentors

wei2912 commented:
Assigning to @JaCh23 for review.

JaCh23 commented:
> As mentioned above, I updated the query to use DuckDB SQL it now will group by mentor_name as the id field will create 2 different rows for some mentors

Much rather we look into this deeper, we should try to work and join by ID for data integrity purposes, else we may run into issues now or later on (eg. 2 diff mentors with same name then accidentally grouped together); can revise query to use id?

vanesssalai commented:
Updated the sql query to group by mentor data. Here are the screenshots
![Screenshot 2025-01-13 201443](https://github.com/user-attachments/assets/cb170cfa-b257-4570-b63b-fc374c324e47)
![Screenshot 2025-01-13 201451](https://github.com/user-attachments/assets/843d6b98-2219-4ae5-98cd-1e076eca8c78)

JaCh23 commented:
LGTM

JaCh23 reviewed (approved):
LGTM

wei2912 commented on dashboards/clicks_and_impressions.py:
```diff
@@ -0,0 +1,92 @@
+cur = get_dbcur()
+
+query = """
+WITH ranked_events AS (
```
Just wondering - why is this named `ranked_events`? Seems like the ranking is only done in [L45-45](https://github.com/AdvisorySG/mentorship-streamlit/pull/46/files#diff-70d43095056fd0c892b4c6ca5c1c4f1882de4af2645b15300f724d302f549120R44-R45).

vanesssalai replied:
It was from my initial try I never renamed after that. will rename to production_events

wei2912 commented on dashboards/clicks_and_impressions.py:
```diff
@@ -0,0 +1,92 @@
+        ed.website_event_id,
+        es.id AS mentor_id,
+        ANY_VALUE(es.name) AS mentor_name,
+        SUM(CASE WHEN we.event_name = 'Click' THEN 1 ELSE 0 END) AS click_count,
+        SUM(CASE WHEN we.event_name = 'Impression' THEN 1 ELSE 0 END) AS impression_count
```
An additional condition is required to remove tracked events on development instances (see https://github.com/AdvisorySG/mentorship-streamlit/pull/46#pullrequestreview-2522572346 for details). Perhaps add something like `SUM(CASE WHEN ed.data_key = 'env' AND ed.string_value = 'production' THEN 1 ELSE 0 END) AS is_production`, then add a filter later on to select only columns with `is_production = 1`?

wei2912 replied:
Actually, such a filter won't work - I see that the join criteria at [L25](https://github.com/AdvisorySG/mentorship-streamlit/pull/46/files#diff-70d43095056fd0c892b4c6ca5c1c4f1882de4af2645b15300f724d302f549120R25) means all of the rows matching `ed.data_key = 'id'` won't be included in this final table. From my understanding, we probably need to do a subquery to retain only the production events in `ed`.

wei2912 commented on dashboards/clicks_and_impressions.py:
```diff
@@ -0,0 +1,92 @@
+        ON es.id = ed.string_value
+    WHERE 
+        (we.event_name = 'Click' OR we.event_name = 'Impression')
+        AND we.url_path LIKE '/mentors%'
+        AND we.referrer_path LIKE '/mentors%'
```
These two criteria shouldn't be necessary, as the click/impression events can only occur on `/mentors%`. (Also, note that `referrer_path` is the previous URL visited by the user, so it should have no bearing on the tracking of click/impression events.)

See the following query which is empty:
![image](https://github.com/user-attachments/assets/76634754-dff6-43cb-9afc-1d1b01c8364a)

wei2912 commented on dashboards/clicks_and_impressions.py:
```diff
@@ -0,0 +1,92 @@
+        ed.website_event_id, 
+        es.id
+)
+
```
From my understanding, each row of `ranked_events` should correspond to a single row in `event_data` with `ed.data_key = 'id'`, as well a single row in `website_event` and `elasticsearch` each since they have primary keys `event_id` and `id` respectively. (Technically, `elasticsearch` doesn't have a primary key since it's from JSON data, but `id` should be unique...)

If this is correct, could we run a check that `click_count + impression_count = 1` (or XOR) for each row?

wei2912 reviewed (changes requested):
Plots portion LGTM, but the SQL query needs some changes. Please let me know if you have any questions - a fair bit of it is complicated :")

vanesssalai commented:
- updated the naming of ranked_events to mentor_events
- added production_events to filter out non production events
- added L24-L27 to check for click_count + impression_count != 1
